prog/weaver/metrics.go

```diff
@@ -0,0 +1,143 @@
+	terminationDesc     *prometheus.Desc
+	flowsDesc           *prometheus.Desc
+	bytesDesc           *prometheus.Desc
+	packetsDesc         *prometheus.Desc
+	ipsDesc             *prometheus.Desc
+	dnsDesc             *prometheus.Desc
+}
+
+func newMetrics(router *weave.NetworkRouter, allocator *ipam.Allocator, ns *nameserver.Nameserver) *metrics {
+	return &metrics{
+		router:    router,
+		allocator: allocator,
+		ns:        ns,
+		connectionCountDesc: prometheus.NewDesc(
+			"weave_connections",
+			"Number of peer-to-peer connections.",
+			[]string{"state"},
+			prometheus.Labels{},
+		),
+		terminationDesc: prometheus.NewDesc(
+			"weave_connection_termination_count",
+			"Number of peer-to-peer connections terminated.",
+			[]string{},
+			prometheus.Labels{},
+		),
+		flowsDesc: prometheus.NewDesc(
+			"weave_flows",
+			"Number of FastDP flows.",
+			[]string{},
+			prometheus.Labels{},
+		),
+		packetsDesc: prometheus.NewDesc(
+			"weave_packets_total",
+			"Number of packets transferred.",
+			[]string{"flow"},
+			prometheus.Labels{},
+		),
+		bytesDesc: prometheus.NewDesc(
+			"weave_bytes_total",
+			"Number of bytes transferred.",
+			[]string{"flow"},
+			prometheus.Labels{},
+		),
+		ipsDesc: prometheus.NewDesc(
+			"weave_ips",
+			"Number of IP addresses.",
+			[]string{"state"},
+			prometheus.Labels{},
+		),
+		dnsDesc: prometheus.NewDesc(
+			"weave_dns_entries",
+			"Number of DNS entries.",
+			[]string{"state"},
+			prometheus.Labels{},
+		),
+	}
+}
+
+func (m *metrics) Collect(ch chan<- prometheus.Metric) {
+	intMetric := func(desc *prometheus.Desc, val int, labels ...string) {
+		ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, float64(val), labels...)
+	}
+	uint64Counter := func(desc *prometheus.Desc, val uint64, labels ...string) {
+		ch <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, float64(val), labels...)
+	}
+
+	routerStatus := weave.NewNetworkRouterStatus(m.router)
+
+	established := 0
+	for _, conn := range routerStatus.Connections {
+		if conn.State == "established" {
+			established++
+		}
+	}
+
+	intMetric(m.connectionCountDesc, len(routerStatus.Connections)-established, "non-established")
+	intMetric(m.connectionCountDesc, established, "established")
+	uint64Counter(m.terminationDesc, uint64(routerStatus.TerminationCount))
+
+	flows := 0
+	var totalPackets, totalBytes uint64
+	if diagMap, ok := routerStatus.OverlayDiagnostics.(map[string]interface{}); ok {
+		if fastDPEntry, ok := diagMap["fastdp"]; ok {
+			if fastDPStatus, ok := fastDPEntry.(weave.FastDPStatus); ok {
+				flows = len(fastDPStatus.Flows)
+				for _, flow := range fastDPStatus.Flows {
+					//tag := flowTag(flow)
+					//uint64Counter(m.packetsDesc, flow.Packets, tag)
+					//uint64Counter(m.bytesDesc, flow.Bytes, tag)
+					totalPackets += flow.Packets
+					totalBytes += flow.Bytes
+				}
+			}
+		}
+	}
+	intMetric(m.flowsDesc, flows)
+	uint64Counter(m.bytesDesc, totalBytes, "total")
```

Comment:
Prometheus [doc](https://prometheus.io/docs/concepts/metric_types/#counter) says that the counter metric cannot be used to represent a value which can go down. At the same time, we periodically flush flows (which includes stats related to them) which means it is possible `totalBytes[now-1]` to be greater than `totalBytes[now]`.